bellman/src/domain/mod.rs

```diff
@@ -0,0 +1,342 @@
+pub mod arith;
+pub mod linear;
+pub mod fft;
+pub mod multiexp;
+
+pub use arith::*;
+pub use linear::*;
+pub use fft::*;
+pub use multiexp::*;
+
+use ff::{Field, PrimeField, ScalarEngine};
+
+use crate::{error, multicore};
+use error::{SynthesisError, Result};
+use multicore::Worker;
+use arith::{Scalar, Group};
+
+use std::ops;
+
+/// A `Domain` abstraction for
+/// performing various kinds of polynomial arithmetic on top of
+/// the scalar field.
+///
+/// In pairing-based SNARKs like Groth16, we need to calculate
+/// a quotient polynomial over a target polynomial with roots
+/// at distinct points associated with each constraint of the
+/// constraint system. In order to be efficient, we choose these
+/// roots to be the powers of a 2^n root of unity in the field.
+/// This allows us to perform polynomial operations in O(n)
+/// by performing an O(n log n) FFT over such a domain.
+pub struct Domain<'a,E,G> 
+where
+    E: ScalarEngine
+{
+    coeffs: Vec<G>,
+    exp: u32,
+    omega: E::Fr,
+    omegainv: E::Fr,
+    geninv: E::Fr,
+    minv: E::Fr,
+    worker: &'a Worker
+}
+
+impl<'a,E,G> Domain<'a,E,G> 
+where
+    E: ScalarEngine,
+    for <'b> G: Group<'b,E>
+{
+    pub fn new(mut coeffs: Vec<G>, worker: &'a Worker) -> Result<Self> {
+        let (m,exp): (usize,u32) = Self::size_of(&coeffs)?;
+        let omega: E::Fr = Self::square_primitive_root_of_unity_to_degree(exp);
+
+        let omegainv: _ = omega.inverse()?;
+        let geninv: _ = E::Fr::multiplicative_generator().inverse()?;
+
+        let casted_m: _ = format!("{}",m);
+        let minv: _ = E::Fr::from_str(&casted_m)?.inverse()?;
+
+        Self::maybe_pad_with_zeroes(&mut coeffs, m);
+
+        let domain: _ = Domain {
+            coeffs,
+            exp,
+            omega,
+            omegainv,
+            geninv,
+            minv,
+            worker
+        };
+        Ok(domain)
+    }
+
+    fn maybe_pad_with_zeroes(coeffs: &mut Vec<G>, size: usize) {
```

Comment:
Don't really love this function name, if tries to pad with zeros every time

Reply:
I’m fairly confident that the .resize() call inside the function will only sometimes add zeroes, hence the name. I’m hesitant to call it try_pad because that implies there’s an error to catch.

Reply:
Ahh. Agreed for now, but I still don't like how maybe for me has conditionals associated (like if, elif, else)

Reply:
What do you mean by has conditionals associated? I will confirm but I believe that resize has an if block inside it,  would that make maybe an appropriate name?

Reply:
Actioned